### ISSUE TITLE
MueLu: what to do when rowMap does not match ColMap?

### ISSUE THREAD

**Issue by aprokop** [trilinos/Trilinos#1747]
@trilinos/muelu 

We have a lot of code that is sensitive to this question, and from I can see the code is buggy:
- CoalesceDropFactory
- FilteredAFactory (noticed in #1678)
- Aggregation

I think MueLu would be a lot better if we *require* that row and column maps match. Otherwise, if we don't, we'll have to check for that everywhere (that is minor), and use error prone othwerise. This will most likely not be noticed until somebody hits a problem with custom maps that we'll take a long time to debug.

Unfortunately, copying a matrix into the format satisfying the requirement is going to be expensive.

**Comment by aprokop** [comment]
Also affected: Ifpack2 preconditioners, for instance, Jacobi and Gauss-Seidel, that rely on getting a correct diagonal, all ILU preconditioners.
It also seems that some of the Tpetra routines implicitly rely on this fact, for instance, `getLocalDiagOffsets` in Graph does not do any global indices check, so in fact may return correct mathematical matrix diagonal offsets.

**Comment by aprokop** [comment]
@mhoemmen I'd like to solicit your opinion on this. How hard do you think would be for preconditioners in Trilinos to require things from applications regarding their column maps?

**Comment by tawiesn** [comment]
@aprokop Not sure whether i understand what you mean: given the fact that our row maps are always supposed to be non-overlapping and therefore the same as the range maps and that the column maps are always overlapping (in parallel), what exactly do you mean? Do you have an example?

**Comment by mhoemmen** [comment]
@aprokop wrote:
> It also seems that some of the Tpetra routines implicitly rely on this fact, for instance, getLocalDiagOffsets in Graph does not do any global indices check, so in fact may return correct mathematical matrix diagonal offsets.

Which routines are you looking at?  `Tpetra::CrsGraph::getLocalDiagOffsets` definitely uses the row and column Maps.  Even the thread-parallel version uses the LocalMap objects.  See `Tpetra_Details_getGraphDiagOffsets_def.hpp`.  In fact, this could use optimizations for the "is fitted" case.

**Comment by mhoemmen** [comment]
@aprokop wrote:
> I'd like to solicit your opinion on this. How hard do you think would be for preconditioners in Trilinos to require things from applications regarding their column maps?

It could be harder than one thinks.  I tried fixing `Ifpack2::LocalFilter` a few years ago to remove the implicit "is fitted" assumption.  It was hard, because `Ifpack2::AdditiveSchwarz`'s implementation of overlap had an unexpected coupling to LocalFilter.  I gave up at the time since there wasn't an application driver.

**Comment by mhoemmen** [comment]
@tawiesn wrote:
> Not sure whether i understand what you mean: given the fact that our row maps are always supposed to be non-overlapping and therefore the same as the range maps and that the column maps are always overlapping (in parallel), what exactly do you mean? Do you have an example?

See the documentation of the `isLocallyFitted` predicate in `Tpetra_Map_decl.hpp`.

**Comment by aprokop** [comment]
> Which routines are you looking at? Tpetra::CrsGraph::getLocalDiagOffsets definitely uses the row and column Maps.

@mhoemmen If I look at the `Tpetra::CrsGraph::getLocalDiagOffsets`, I see this:
```c++
    // mfh 12 Mar 2016: LocalMap works on (CUDA) device.  It has just
    // the subset of Map functionality that we need below.
    auto lclRowMap = rowMap.getLocalMap ();
    auto lclColMap = colMap.getLocalMap ();

    // FIXME (mfh 16 Dec 2015) It's easy to thread-parallelize this
    // setup, at least on the host.  For CUDA, we have to use LocalMap
    // (that comes from each of the two Maps).

    const bool sorted = this->isSorted ();
    if (isFillComplete ()) {
      auto lclGraph = this->getLocalGraph ();
      Details::getGraphDiagOffsets (offsets, lclRowMap, lclColMap,
                                    lclGraph.row_map, lclGraph.entries,
                                    sorted);
```
So, if a matrix is fill complete, it immediately goes to local orderings. It does *not* check for whether the maps are locally fitted, and thus may return something weird. I hope I get it right.

**Comment by aprokop** [comment]
> It could be harder than one thinks. I tried fixing Ifpack2::LocalFilter a few years ago to remove the implicit "is fitted" assumption. It was hard, because Ifpack2::AdditiveSchwarz's implementation of overlap had an unexpected coupling to LocalFilter. I gave up at the time since there wasn't an application driver.

No, I think you misunderstood me. I'm not arguing for removing the assumption for maps to be locally fitted. Instead, I'm arguing for making it mandatory to be able to use preconditioners.

**Comment by aprokop** [comment]
> @aprokop Not sure whether i understand what you mean: given the fact that our row maps are always supposed to be non-overlapping and therefore the same as the range maps and that the column maps are always overlapping (in parallel), what exactly do you mean? Do you have an example?

@tawiesn I'm talking about being locally fit. Specifically, column_gids = [row_gids, exras]. As Mark pointed out, `isLocallyFitted` documentation is a better description. For us, this was call `isGoodMap` in the TentativeP factory. Essentially, it means that GIDs corresponding to local indices in row map are the same as GIDs in the column map with the same local indices.

**Comment by lucbv** [comment]
@aprokop isLocallyFitted does not require two maps to be the same at all.
If a `rowMap` contains only on rank entries and the `colMap` contains a superset of said on rank entries plus off rank entries.
Most of the time `isLocallyFitted(rowMap,colMap)` will return true unless your rowMap contains off rank entries which seems unlikely.

**Comment by mhoemmen** [comment]
@aprokop wrote:
> No, I think you misunderstood me. I'm not arguing for removing the assumption for maps to be locally fitted. Instead, I'm arguing for making it mandatory to be able to use preconditioners.

Ah, sorry, I misunderstood you there.  I think it's reasonable for now to require this, at least until we have time to find all the places that make this implicit assumption.

**Comment by aprokop** [comment]
> @aprokop isLocallyFitted does not require two maps to be the same at all.

@lucbv Yes, I understand that. And when running in parallel, the are usually not the same (row map and column map).

> If a rowMap contains only on rank entries and the colMap contains a superset of said on rank entries plus off rank entries then rowMap is locally fitted.

This is the exact point of the discussion. This is only true if the local entries for the column map are in the same order as those of the row map. If not, then they are not locally fitted.

**Comment by mhoemmen** [comment]
@aprokop wrote:
> So, if a matrix is fill complete, it immediately goes to local orderings. It does not check for whether the maps are locally fitted, and thus may return something weird. I hope I get it right.

Check out `Tpetra_Details_getGraphDiagOffsets_def.hpp`, lihes 109-111.  The kernel converts from local row index, to global index, to local column index, using the (local parts of the) row and column Maps.

**Comment by lucbv** [comment]
@mhoemmen looking at `isLocallyFitted`, it seems that the check for `numInds_map1 > inds_map2.size ()` can be made prior to checking `map1.isContiguous ()`, it could save a little bit of code and time...

**Comment by aprokop** [comment]
@mhoemmen You are right, I was staring at the code, but somehow missed that 😬 So Tpetra's off the hook :) Though it also means that getting the diagonal is quite expensive (and I saw that in the timers) as it does GID -> LID translations.

**Comment by aprokop** [comment]
@lucbv In #1714 we reorganize locally fitted and add some improvements, including the one you mentioned.

**Comment by lucbv** [comment]
@aprokop I see, I was not aware of the branch so I just read the current version in develop.

**Comment by mhoemmen** [comment]
@aprokop Is #1714 ready for (re)examination?

**Comment by aprokop** [comment]
> @aprokop Is #1714 ready for (re)examination?

@mhoemmen Yes, I think it's ready.

**Comment by jhux2** [comment]
So we propose that MueLu will start enforcing this for user-supplied matrices.  It would also make sense that Ifpack2 also enforce this.  I will bring up this issue at our regular internal applications meeting.

**Comment by rppawlo** [comment]
@eric-c-cyr and @egphill - any impact on Teko?

**Comment by jhux2** [comment]
@aprokop @maherou recommended that this be publicized on the users list prior to any change, including how an application can check whether its matrices satisfy the locally-fitted property.

**Comment by aprokop** [comment]
@jhux2 So, 1) enable a warning in MueLu; 2) publicize to trilinos-users; 3) throw exception? How do we check application compliance prior to 3)?

**Comment by jhux2** [comment]
@aprokop We can check a subset of internal applications, but the rest will have to do their own checking.

**Comment by jhux2** [comment]
What would be a reasonable window of time to let applications respond?

**Comment by aprokop** [comment]
@jhux2 Couple hours? Seriously, though, applications tend to move at glacial speeds on things like that. So, maybe, it's better to reach specific people with a patch, so that they could run their test suits with it, and report back. For that, I would give few weeks.

Hidden behind all this is the fact that I care less about how the old non-Kokkos version runs, and will just enable the throw for the Kokkos version. Once we get feedback from apps, we can switch it on for non-Kokkos too.

**Comment by jhux2** [comment]
It just occurred to me, do the Epetra-based solvers make any similar assumptions?

**Comment by mhoemmen** [comment]
@jhux2 AztecOO and ML almost certainly do.  I don't know if they check input at all for this condition.

**Comment by csiefer2** [comment]
By "row maps and column maps match" do you mean that "all entries in the row map are numbered with the same GID/LID combo as the column map"  (additional column map entries come later).

@jhux2 I believe that Ifpack requires the above condition sometimes (but is inconsistent about checking it).

**Comment by mhoemmen** [comment]
@csiefer2 wrote:
> By "row maps and column maps match" do you mean that "all entries in the row map are numbered with the same GID/LID combo as the column map" (additional column map entries come later).

Yes.  It means "the column Map is locally fitted to the row Map, on all processes."

> I believe that Ifpack requires the above condition sometimes....

LocalFilter _definitely_ does.  LocalFilter is tied to AdditiveSchwarz overlap in a way that makes this nontrivial to fix.

**Comment by csiefer2** [comment]
This would be something trivial to forcibly check in a DEBUG build (with the associated reduction) and not in a RELEASE one.

This would allow us to "require" local fitting without having to check in a Release build

**Comment by csiefer2** [comment]
One caveat is that this might cause some degree of havoc with Xpetra::BlockedCrsMatrices, if you try to do something like get the diagonal of a square block off the diagonal...

**Comment by tawiesn** [comment]
> One caveat is that this might cause some degree of havoc with Xpetra::BlockedCrsMatrices, if you try to do something like get the diagonal of a square block off the diagonal...

Where would you need to extract the diagonal of an off-diagonal block in a blocked operator? In general the off-diagonal blocks in a blocked operator have rectangular shape so it's not even clear what the diagonal would be. Don't think that is a real issue.

All block smoothers in MueLu so far only need proper blocks on the diagonal of the blocked operator. They need to be locally fitted as we call other smoothers (e.g. from Ifpack) on them. The off-diagonal blocks are only needed to calculate the residual or rhs within the block smoothers.

**Comment by csiefer2** [comment]
@tawiesn I never said it was a *good* idea, but if you have no official concerns, then I don't either...

**Comment by tawiesn** [comment]
@csiefer2 We might want/need to add some checks to the BlockedCrsMatrix to help the user avoid unnecessary mistakes/problems. I just meant that i don't see immediate problems when enforcing locally fitted Ids. The only thing we need is a MV multiplication

**Comment by mhoemmen** [comment]
@csiefer2 wrote:
> This would be something trivial to forcibly check in a DEBUG build (with the associated reduction) and not in a RELEASE one.

Not all preconditioners care: For example, (point) Jacobi and Gauss-Seidel can extract the matrix diagonal without needing this condition.  This suggests that only the preconditioners that care should check.

**Comment by eric-c-cyr** [comment]
@tawiesn @csiefer2 I could see wanting the lumped action or diagonal of the off diagonal block. I'm not sure what this would mean in a  non-rectangular case. Probably you would just lump the operator rather than take the diagonal. I feel certain that this is  case one might actually consider (though I'm not sure we have considered it to date).

**Comment by aprokop** [comment]
> Not all preconditioners care: For example, (point) Jacobi and Gauss-Seidel can extract the matrix diagonal without needing this condition. This suggests that only the preconditioners that care should check.

@mhoemmen Yes, I think it is correct to say that only *some* Ifpack2 preconditioners are affected by the implicit assumption. So, I think we should add this check only to the ones we *know*.

**Comment by jhux2** [comment]
> So, 1) enable a warning in MueLu; 2) publicize to trilinos-users; 3) throw exception? How do we check application compliance prior to 3)?

@aprokop There were no objections raised when I proposed this on the Trilinos Users list.

**Comment by aprokop** [comment]
OK, I already forgot about this one. So we simply enforce it now and check the matrix user provided?

**Comment by jhux2** [comment]
Yes, fine with me.

**Comment by mhoemmen** [comment]
@csiefer2 is this still a thing?
